### PR TITLE
Remove TimeExtractor and use internal API for setting start/end timestamps

### LINKED ISSUE #6014: Some Instrumentation API thoughts
* `VirtualField` - maybe move to `util` package, just to avoid a single class package?
* `RequestListener` and `RequestMetrics` - maybe rename to `OperationListener` and `OperationMetrics`?
  * The terms Request/Response I think are good when used to mean the request object / response object, but I think "Request" in these names is used more generally to mean the "request-response-loop".
* `SpanStatusExtractor` - I was thinking maybe we should rename to `StatusExtractor` because I suspect there will be a semantic convention for capturing status on metrics as well (https://github.com/open-telemetry/opentelemetry-specification/issues/2419#issuecomment-1081923604), but `StatusCode` enum is under `io.opentelemetry.api.trace`, so the future is not so clear, and maybe we should stay as-is, and we can always add a common `StatusExtractor` in the future if we want.
* `SpanLinksExtractor.fromUpstreamRequest()` - maybe rename to `fromTextMapPropagator()` and pass in `TextMapPropagator` instead of `ContextPropagators`? This would also make it future-proof if/when "Binary" Propagator is introduced.
* `TimeExtractor` - maybe slim it down to just `StartTimeExtractor`? this would satisfy all of our use cases, and make it clearer that overriding end time doesn't really work for metrics anyways

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Replaces https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6024, resolves #6014

### CODE REVIEW

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/InstrumenterUtil.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      Method method =
+          Instrumenter.class.getDeclaredMethod(
+              "startAndEnd",
+              Context.class,
+              Object.class,
+              Object.class,
+              Throwable.class,
+              long.class,
+              long.class);
+      method.setAccessible(true);
```
Comment (trask) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/InstrumenterUtil.java:43]: can we cache this?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java [hunk 2]
```diff
@@ -129,17 +126,47 @@ public boolean shouldStart(Context parentContext, REQUEST request) {
...
+  }
+
+  /** Internal method for creating spans with given start/end timestamps. */
+  Context startAndEnd(
+      Context parentContext,
+      REQUEST request,
+      @Nullable RESPONSE response,
+      @Nullable Throwable error,
+      long startTimeNanos,
+      long endTimeNanos) {
```
Comment (trask) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java:154]: curious your thoughts on:

```suggestion
      Instant startTime,
      Instant endTime) {
```

**Review by trask — APPROVED** [review]

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java (hunk 2 above)
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java]: 👍

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/InstrumenterUtil.java (hunk 1 above)
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/InstrumenterUtil.java]: Definitely 👍

**Review by laurit — APPROVED** [review]
These spans won't always be in correct order when compared to other spans when running on jdk8 because `Instant.now` only has millisecond resolution in jdk8.

**Comment by mateuszrzeszutek** [comment]
> These spans won't always be in correct order when compared to other spans when running on jdk8 because `Instant.now` only has millisecond resolution in jdk8.

Yep -- and they also may be out of order because they're not using the same `AnchoredClock` as the other spans. That's one of the reasons we're making it internal, I guess.

**Review by trask — APPROVED** [review]

**Pull request merged**